src/main/kotlin/org/rust/lang/core/types/infer/TypeInference.kt

```diff
@@ -1124,12 +1124,20 @@ class RsFnInferenceContext(
     }
 
     private fun pickSingleMethod(receiver: Ty, variants: List<MethodResolveVariant>, methodCall: RsMethodCall): MethodResolveVariant? {
-        val filtered = variants.singleOrFilter {
+        val filtered = variants.singleOrLet { list ->
             // 1. filter traits that are not imported
             TypeInferenceMarks.methodPickTraitScope.hit()
-            val trait = it.source.impl?.traitRef?.path?.reference?.resolve() as? RsTraitItem
-                ?: return@singleOrFilter true
-            lookup.isTraitVisibleFrom(trait, methodCall)
+            val traitToCallee = hashMapOf<RsTraitItem, MutableList<MethodResolveVariant>>()
+            val inherent = mutableListOf<MethodResolveVariant>()
+            for (callee in list) {
+                val trait = callee.source.impl?.traitRef?.resolveToTrait
+                if (trait != null) {
+                    traitToCallee.getOrPut(trait) { mutableListOf() }.add(callee)
+                } else {
+                    inherent.add(callee)
+                }
+            }
+            inherent + traitToCallee.keys.filterInScope(methodCall).map(traitToCallee::getValue).flatten()
```

Comment:
It seems the following implementation is little more efficient
```kotlin
traitToCallee.keys.filterInScope(methodCall).forEach {
    inherent += traitToCallee.getValue(it)
}
inherent
```
But of course, in this case we should rename `inherent` list